image.py

```diff
@@ -0,0 +1,355 @@
+import io
+from shlex import split
+from subprocess import CalledProcessError, run
+from sys import stdin
+from typing import Union
+
+from exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
+    """
+
+    def __init__(self, name_or_id: str):
+        """Initialize this Image. Connects the class with its ID.
+
+        Parameters
+        ----------
+        name_or_id : str
+            A name or ID by which to find this image using docker inspect.
+        """
+        self._id = get_image_id(name_or_id)
+
+    @classmethod
+    def build_from_dockerfile(
+        cls,
+        context: str,
+        tag: str,
+        output_file: Union[io.TextIOBase, None] = None,
+        dockerfile_loc: str = ""
```

Comment:
This suggestion is sort of a matter of personal preference, so feel free to ignore.

Personally, I would name this argument as follows to try and cut down on verbosity a bit without sacrificing clarity (IMO):

```suggestion
        dockerfile: str = ""
```

Reply:
Like the `context` argument above, the type of `dockerfile` should be `os.PathLike[str]` to be more flexible.

```suggestion
        dockerfile: os.PathLike[str] = ""
```

Reply:
Let's change the default value to `None` instead of an empty string. 

```suggestion
        dockerfile: Optional[os.PathLike[str]] = None
```

Then, if the implementation wants it to be a string, you could have something like this in the function body:

```python
if dockerfile is None:
    dockerfile = ""
```

Among other reasons, a `str` is a mutable type so we should prefer not to use an empty string as a default value when possible in order to avoid falling into the common pitfall of using [mutable default arguments](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments). 

In general, I think the dangers of using mutable default arguments are somewhat overblown and in many cases I'm fine with using a `str` as a default value. But in this case, using `None` feels more idiomatic to me anyway.